### PR TITLE
Port some changes from arangod, June 2019

### PR DESCRIPTION
**Pull request by dhly-etc**

### CODE REVIEW

File: include/fuerte/detail/vst.h [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  
+namespace message {
+  
+/// @brief creates a slice containing a VST request-message header.
+void requestHeader(RequestHeader const&, velocypack::Buffer<uint8_t>&);
+/// @brief creates a slice containing a VST response-message header.
+void responseHeader(ResponseHeader const&, velocypack::Buffer<uint8_t>&);
+/// @brief creates a slice containing a VST auth message with JWT encryption
+void authJWT(std::string const& token, velocypack::Buffer<uint8_t>&);
+/// @brief creates a slice containing a VST auth message with plain enctyption
```
Comment (jsteemann) [inline include/fuerte/detail/vst.h]:
```suggestion
/// @brief creates a slice containing a VST auth message with plain encryption
```

File: include/fuerte/detail/vst.h [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  // This function is only valid for first chunks.
+  inline uint32_t numberOfChunks() const {
+    if (isFirst()) {
+      return _chunkX >> 1;
+    }
+    assert(false);  // illegal call
+    return 0;       // Not known
+  }
+
+  // writeHeaderToVST1_0 write the chunk to the given buffer in VST 1.0 format.
```
Comment (jsteemann) [inline include/fuerte/detail/vst.h]:
```suggestion
  // writeHeaderToVST1_0 writes the chunk to the given buffer in VST 1.0 format.
```

File: include/fuerte/detail/vst.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  // The length of the buffer is returned.
+  size_t writeHeaderToVST1_0(size_t chunkDataLen, velocypack::Buffer<uint8_t>&) const;
+
+  // writeHeaderToVST1_1 write the chunk to the given buffer in VST 1.1 format.
```
Comment (jsteemann) [inline include/fuerte/detail/vst.h]:
```suggestion
  // writeHeaderToVST1_1 writes the chunk to the given buffer in VST 1.1 format.
```

File: include/fuerte/detail/vst.h [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+/// @brief creates a slice containing a VST response-message header.
+void responseHeader(ResponseHeader const&, velocypack::Buffer<uint8_t>&);
+/// @brief creates a slice containing a VST auth message with JWT encryption
+void authJWT(std::string const& token, velocypack::Buffer<uint8_t>&);
+/// @brief creates a slice containing a VST auth message with plain enctyption
+void authBasic(std::string const& username,
+               std::string const& password,
+               velocypack::Buffer<uint8_t>&);
+
+/// @brief take existing buffers and partitions into chunks
```
Comment (jsteemann) [inline include/fuerte/detail/vst.h:104]:
```suggestion
/// @brief take existing buffer and partition it into chunks
```

File: src/MessageStore.h [hunk 5]
```diff
@@ -35,19 +35,17 @@
 
 namespace arangodb { namespace fuerte { inline namespace v1 {
 // MessageStore keeps a thread safe list of all requests that are "in-flight".
-template <class RequestItemT>
+template <typename RequestItemT>
 class MessageStore {
  public:
   // add a given item to the store (indexed by its ID).
   void add(std::shared_ptr<RequestItemT> item) {
-    std::lock_guard<std::mutex> lockMap(_mutex);
```
Comment (jsteemann) [inline src/MessageStore.h:43]: as this mutex is removed here entirely, I hope there is now some other external synchronization for accessing the `MessageStore`.

File: src/vst.cpp [hunk 6]
```diff
@@ -159,14 +161,12 @@ VPackBuffer<uint8_t> message::requestHeader(RequestHeader const& header) {
...
-
-  return buffer;
 }
 
-/// @brief creates a slice containing a VST request header.
-VPackBuffer<uint8_t> message::responseHeader(ResponseHeader const& header) {
+/// @brief creates a slice containing a VST response-message header.
+void message::responseHeader(ResponseHeader const& header,
+                             VPackBuffer<uint8_t>& buffer) {
   static std::string const message = " for message not set";
```
Comment (jsteemann) [inline src/vst.cpp]: this static string seems completely unnecessary. it seems to be used only once below, and is then concatened to another const char sequence. So let's remove it here and adjust line 188 to throw the complete message.

File: src/vst.cpp [hunk 7]
```diff
@@ -226,51 +222,38 @@ VPackBuffer<uint8_t> message::authBasic(std::string const& username,
...
+                              std::vector<asio_ns::const_buffer>& result) {
   // Split message into chunks
 
-  size_t msgLength = payload.size() + _requestMetadata.size();
+  // we assume that the message header is already in the buffer
+  size_t msgLength = buffer.size() + payload.size();
   assert(msgLength > 0);
 
-  // builds a list of chunks that are ready to be send to the server.
+  // builds a list of chunks that are ready to beheader send to the server.
```
Comment (jsteemann) [inline src/vst.cpp]:
```suggestion
  // builds a list of chunks that are ready to be sent to the server.
```

**Review by jsteemann — CHANGES_REQUESTED** [review]

File: src/MessageStore.h (hunk 5 above)
Comment (dhly-etc) [reply src/MessageStore.h:43]: So from my understanding of @graetzer's explanation, the `MessageStore` is only accessed by the single `asio` thread, and therefore does not need to be synchronized. All the locking code had previously been commented out for the better part of the last year in arangod, and I just removed it when porting changes.

Comment (jsteemann) [reply src/MessageStore.h:43]: Ok, alright then.

File: src/VstConnection.cpp [hunk 8]
```diff
@@ -401,7 +400,7 @@ void VstConnection<ST>::asyncWriteCallback(asio_ns::error_code const& ec,
   FUERTE_LOG_CALLBACKS << "asyncWriteCallback (vst): send succeeded, "
                        << transferred << " bytes transferred\n";
 
-  // request is written we no longer need data for that
+  // request is written we no longer headerneed data for that
```
Comment (jsteemann) [inline src/VstConnection.cpp]:
```suggestion
  // request is written we no longer need data for that
```

**Review by jsteemann — CHANGES_REQUESTED** [review]

**Review by graetzer — APPROVED** [review]

**Review by jsteemann — APPROVED** [review]

**Pull request merged**